### ISSUE TITLE
Plot: Create GetDataLimits() for returning X/Y limits of all plotted data

### ISSUE THREAD

**Issue by EFeru** [ScottPlot/ScottPlot#1415]
**ScottPlot Version:** 4.1.26

**Operating System:** W10

**Application Type:** WinForms

**Question:** I need to limit zoom out. I would like that it doesn't go beyond the AxisAuto(). I know I can implement on application side something to compute Min/Max of existing data and use `SetOuterViewLimits` to achieve this. However, I am wondering, wouldn't be easier and more suited to have this functionality build in into Scottplot?
It could be a flag that we can set to limit zoom out to maximum the AxisAuto() fit.

**Comment by swharden** [comment]
Hi @EFeru, thanks for posting this issue! I agree the current strategy is a bit clunky, and I'd love to figure-out a way to improve it.

Calculating min/max of existing data is actually pretty costly performance wise, so I wouldn't want to do it automatically before each render. Instead I think the way to go is to have the user perform this task (they know best when the data may change).

What do you think about code that did something like this?

```cs
AxisLimits limits = myPlot.GetDataLimits();
myPlot.SetOuterViewLimits(limits);
```

Basically it's a way to get the limits from `AxisAuto()` without actually resizing the window yet. What do you think about this potential solution?

**Comment by EFeru** [comment]
GetDataLimits should do the job. And gives flexibility to the user. I can call it when I know the data is changed. 
I didn't check, but does ScottPlot detects when the data is changing? Because it would be useful for improving speed of AxisAuto, by only computing the limits when a change is detected.

**Comment by bclehmann** [comment]
> I didn't check, but does ScottPlot detects when the data is changing?

No it does not.

**Comment by EFeru** [comment]
Hi @swharden ,

- I did a quick try in my application with the new `GetDataLimits()` and for some reason the limits I get are NaN, and I don't know what could be the reason. Any idea?
![image](https://user-images.githubusercontent.com/24780745/148036267-a8e86c76-856d-4bc4-8ffb-57ba66365faf.png)
I am not sure why, because I definitely have data in my plot.
I tryied making a separate project with only one formsplot and there it works but not in my application which is strange.

- And very minor the `myPlot.SetOuterViewLimits(limits)` doesn't work because `limits` as argument directly is not expected, so I had to replace it with:
```cs
myPlot.SetOuterViewLimits(limits.XMin, limits.XMax, limits.YMin, limits.YMax);
```

**Comment by swharden** [comment]
Hi @EFeru, 
Could the problem be that your dataset contains `double.NaN`, or that you have an empty plot?

this .NET console app works for me...

```cs
ScottPlot.Plot plt = new(400, 300);
Random rand = new(0);
double[] xs = ScottPlot.DataGen.Random(rand, 10);
double[] ys = ScottPlot.DataGen.Random(rand, 10);
plt.AddScatter(xs, ys);
Console.WriteLine(plt.GetDataLimits());
plt.SaveFig("test.png");
```

```
AxisLimits: x=[0.2060331540210327, 0.9775497531413798] y=[0.030366990729406004, 0.9953470812157481]
```

![test](https://user-images.githubusercontent.com/4165489/148143342-a15afbaf-18bb-4fc6-80a6-f361e4222a08.png)

**Comment by swharden** [comment]
Oh, I see more recent conversation in #1532...

Restating the problem, adding a plottable with an infinite axis (e.g., axis lines which extend to infinity) results in unexpected behavior:

```cs
ScottPlot.Plot plt = new(400, 300);
Random rand = new(0);
double[] xs = ScottPlot.DataGen.Random(rand, 10);
double[] ys = ScottPlot.DataGen.Random(rand, 10);
plt.AddScatter(xs, ys);
plt.AddVerticalLine(13); // <-- this extends to infinity in the Y direction
Console.WriteLine(plt.GetDataLimits());
```

```
AxisLimits: x=[0.2060331540210327, 13] y=[NaN, NaN]
```

I'll add a comment in #1532 regarding how to best fix this
